sympy/core/parameters.py

```diff
@@ -0,0 +1,127 @@
+"""Thread-safe global parameters"""
+
+from .cache import clear_cache
+from contextlib import contextmanager
+from threading import local
+
+class _global_parameters(local):
+    """
+    Thread-local global parameters.
+
+    Explanation
+    ===========
+
+    This class generates thread-local container for SymPy's global parameters.
+    Every global parameters must be passed as keyword argument when generating
+    its instance.
+    A variable, `global_parameters` is provided as default instance for this class.
+
+    Examples
+    ========
+
+    >>> from sympy.abc import x
+    >>> from sympy.core.cache import clear_cache
+    >>> from sympy.core.parameters import global_parameters as gp
+
+    >>> gp.evaluate
+    True
+    >>> x+x
+    2*x
+
+    Since SymPy's cache is not thread-local, it must be cleared when the thread
+    begins and ends.
```

Comment:
This advice is incorrect. If I have two threads running and one of them starts using `global_evaluate=False` then if both threads share the same cache then each thread can end up receiving incorrect results cached by the other. I think it is best not to mention anything about using threads here apart from the fact that global parameters are stored in thread-local storage. Or perhaps it would just be good to have a warning that although the parameters are stored in thread-local storage the cache can interfere with that so sympy isn't really thread-safe with these parameters.

Reply:
That's a good point. I'll change the docstring.

Reply:
Changed the docstring. It seems that cache really needs to be thread-local.  
After this PR, I will try to fix that.